### DISCUSSION TITLE
How can Maestro reproduce tap gestures and scrolling like a human hand?

### DISCUSSION THREAD

**Discussion by Wayne-Kim** [mobile-dev-inc/Maestro#1729] [Q&A]
How can Maestro reproduce tap, gestures and scrolling like a human hand?

I wonder how this is possible internally.

There is so much code that I really don't know where to start.

Is there anyone who can give me some hints?

**Comment by portyu9** [comment]
Hi @Wayne-Kim 

There are two separate pieces to look at:

1. **Finding the element** — Maestro queries the platform's accessibility hierarchy and resolves the target element/coordinates.
2. **Performing the gesture** — the resolved coordinates/gesture are delegated to the platform-specific driver.

A useful place to start reading the code is `Orchestra.kt`, which dispatches parsed Maestro commands. From there, follow the command into the Maestro client/driver implementation and then into the platform-specific Android/iOS driver.

For example, on Android, the current on-device driver ultimately performs a tap using `UiDevice` (`uiDevice.clickExt(...)`). The higher layers handle things such as selector resolution, waiting/retries and converting an element match into coordinates before the platform driver performs the physical interaction.

So Maestro isn't trying to reconstruct a literal human hand. It uses accessibility information to understand **where** to interact and native/platform automation facilities to inject the tap/swipe/scroll.

I would follow one simple command such as `tapOn` end-to-end first; it is much easier to understand the architecture that way than starting from the repository root and reading every module.

Please mark this answer as accepted if it helped, thank you.
